src/GraftletLoader.cpp

```diff
@@ -131,11 +138,249 @@ bool GraftletLoader::findGraftletsAtDirectory(std::string directory, std::string
+                m_graph.erase(it0);
+                m_dll2ver.erase(name);
+                changed = true;
+                break;
+            }
+        }
+    }
+    return res;
+}
+
+void GraftletLoader::Graph::removeFailedDependants(GraftletLoader& gl)
+{
+    std::vector<DllName> list = removeFailedDependants();
+    for(auto& dllName : list)
+    {
+        gl.m_name2lib.erase(dllName);
+        gl.m_name2registries.erase(dllName);
+    }
+    assert(gl.m_name2lib.size() == gl.m_name2registries.size());
+}
+
+std::string GraftletLoader::Graph::findCycles(bool dont_throw)
+{
+    //find graph cycles
+    // (graph_t::iterator is a map::iterator, the iterators are not bidirectional and have no compare method)
+    auto cmp_graph_iters = [](const graph_t::iterator l, const graph_t::iterator r)->bool { return l->first < r->first; };
+    using vert_clr_t = std::map<graph_t::iterator, Color, decltype(cmp_graph_iters)>;
+    vert_clr_t vert_clr(cmp_graph_iters);
+    //init with white
+    for(auto it = m_graph.begin(), eit = m_graph.end(); it!=eit; ++it)
+    {
+        auto res = vert_clr.try_emplace(it, Color::white);
+        assert(res.second);
+    }
+
+    auto color = [](vert_clr_t::iterator cur_it)->Color& { return cur_it->second; };
+
+    auto find_child = [this, &vert_clr] (DllName& child) -> vert_clr_t::iterator
+    {
+        auto child_iter = m_graph.find(child);
+        assert(child_iter != m_graph.end());
+        auto it = vert_clr.find(child_iter);
+        assert(it != vert_clr.end());
+        return it;
+    };
+
+    while(true)
+    {
+        auto cur_it = vert_clr.begin();
+        //find first white color
+        for(; (cur_it != vert_clr.end()) && color(cur_it) != Color::white; ++cur_it );
+        if(cur_it == vert_clr.end()) break; //not found; really all done
+
+        //almost a stack
+        std::vector< vert_clr_t::iterator > path;
+        color(cur_it) = Color::gray;
+        //depth first search
+        for(;;)
+        {
+            list_t& list = cur_it->first->second;
+            //find white child
+            bool found = false;
+            for(auto& item : list)
+            {
+                DllName& child = item.first;
+                auto child_iter = find_child(child);
+                if(color(child_iter) == Color::gray)
+                {//cycle found
+                    //find the child in the path
+                    auto start = std::find(path.begin(), path.end(), child_iter);
+                    assert(start != path.end() || child_iter == cur_it); //or depends on itself
+                    std::ostringstream oss;
+                    oss << "graftlet dependency cycle found: ";
+                    std::for_each(start, path.end(), [&oss](auto it){ oss << it->first->first << " -> "; } );
+                    oss << cur_it->first->first << " -> " << child;
+                    if(dont_throw) return oss.str();
+                    throw std::runtime_error(oss.str());
```

Comment:
It was intended. For test purposes, there is a possibility to return false instead of throwing an exception. But normally we don't catch and the result will be a stack dump.